apsuite/optics_analysis/optics_correction.py

```diff
@@ -0,0 +1,215 @@
+
+    def _set_delta_kl(self, model=None, knobsidx=None, deltas_kl=None):
+        """Set integrated quadrupoles strengths in the model."""
+        if model is None:
+            model = self.model
+        if knobsidx is None:
+            knobsidx = self.knobs_idx
+        if deltas_kl is None:
+            raise Exception('Missing Delta KL values')
+        for idx_mag, mag in enumerate(knobsidx):
+            delta = deltas_kl[idx_mag]/len(mag)
+            for _, seg in enumerate(mag):
+                stren = model[seg].KL
+                if self.method == OpticsCorr.METHODS.Proportional:
+                    stren *= (1 + delta)
+                else:
+                    stren += delta
+                model[seg].KL = stren
+
+    @staticmethod
+    def get_figm(res):
+        """Calculate figure of merit from residue vector."""
+        return np.sqrt(np.sum(res*res)/res.size)
+
+    def optics_corr_loco(self,
+                         model,
+                         goal_model=None,
+                         jacobian_matrix=None,
+                         nsv=None, nr_max=10, tol=1e-6):
+        """Optics correction LOCO-like.
+
+        Calculates the pseudo-inverse of optics correction matrix via SVD
+        and minimizes the residue vector:
+        Delta [Bx@BPM, Bx@CH, By@BPM, By@CV, Dx@BPM, Tunex, Tuney].
+        """
+        if jacobian_matrix is None:
+            jmat = self.calc_jacobian_matrix(model)
+        else:
+            jmat = _dcopy(jacobian_matrix)
+
+        nominal_stren = self._get_kl(model)
+        if self.method == OpticsCorr.METHODS.Proportional:
+            jmat *= nominal_stren.flatten()
+
+        umat, smat, vmat = np.linalg.svd(jmat, full_matrices=False)
+        ismat = 1/smat
+        ismat[np.isnan(ismat)] = 0
+        ismat[np.isinf(ismat)] = 0
+        if nsv is not None:
+            ismat[nsv:] = 0
+        ismat = np.diag(ismat)
+        ijmat = np.dot(np.dot(vmat.T, ismat), umat.T)
+        goal_vec = self._get_optics_vector(goal_model)
+        vec = self._get_optics_vector(model)
+        dvec = goal_vec - vec
+        bestfigm = OpticsCorr.get_figm(dvec)
+        if bestfigm < tol:
+            return OpticsCorr.CORR_STATUS.Sucess
+
+        for _ in range(nr_max):
+            dkl = np.dot(ijmat, dvec)
+            self._set_delta_kl(model=model, deltas_kl=dkl)
+            vec = self._get_optics_vector(model)
+            dvec = goal_vec - vec
+            figm = OpticsCorr.get_figm(dvec)
+            if figm < bestfigm:
+                bestfigm = figm
+            if bestfigm < tol:
+                break
+        else:
+            return OpticsCorr.CORR_STATUS.Fail
+        return OpticsCorr.CORR_STATUS.Sucess
+
+    def optics_correction(self,
+                          model,
+                          goal_model=None,
+                          jacobian_matrix=None,
+                          nsv=None, nr_max=10, tol=1e-6,
+                          method='loco'):
```

Comment:
I think this `method`option should be an object property, not an input argument.

Reply:
done.

Reply:
see comments below.